File: 7/7.11.swift
/* 
書籍ではファーストクラスコレクションという
コレクションに関する処理をカプセル化したクラスを用意するとあるが
Swiftではそもそもコレクション(配列)に機能を拡張させることができるので
新しいクラスは不要
*/

struct Member {
    var id: Int
    var isAlive: Bool
}

// [GOOD]
// Memberの配列を操作する処理はここに集約される
// 書籍に書いていることをあれこれ考える必要はない
extension [Member] {
    mutating func add(newMember: Member, maxMemberNumber: Int = 4) {
        if self.contains(where: { $0.id == newMember.id }) {
            fatalError("すでに存在している")
        }
        if self.count >= maxMemberNumber {
            fatalError("最大人数を超えています")
        }
        self.append(newMember)
    }
    
    var partyIsAlive: Bool {
        members.contains(where: { $0.isAlive })
    }
}

var members = (1...3).compactMap {
    Member(id: $0, isAlive: true)
}

members.add(newMember: Member(id: 10, isAlive: true))

print(members)

File: 7/7.11.dart
class Member {
  final int id;
  final bool isAlive;

  Member(this.id, this.isAlive);
}

extension MemberList on List<Member> {
  // [GOOD]
  // Memberの配列を操作する処理はここに集約される
  // 書籍に書いていることをあれこれ考える必要はない
  void addMember(Member newMember, {int maxMemberNumber = 4}) {
    if (this.any((member) => member.id == newMember.id)) {
      throw Exception("すでに存在している");
    }
    if (this.length >= maxMemberNumber) {
      throw Exception("最大人数を超えています");
    }
    this.add(newMember);
  }
  
  bool get partyIsAlive => this.any((member) => member.isAlive);
}

void main() {
  final members = List<Member>.generate(3, (index) => Member(index + 1, true));
  members.addMember(Member(10, true));
  print(members);
}
